crates/node-builder/src/exex.rs

```diff
@@ -0,0 +1,96 @@
+#![allow(dead_code)]
+
+use crate::FullNodeTypes;
+use futures::{future::BoxFuture, Future, FutureExt, Stream};
+use reth_node_core::{
+    dirs::{ChainPath, DataDirPath},
+    node_config::NodeConfig,
+};
+use reth_primitives::{BlockNumber, Head};
+use reth_tasks::TaskExecutor;
+use std::{borrow::Cow, pin::Pin};
+
+/// An ExEx (Execution Extension) that processes new blocks and emits events on a stream.
+pub trait ExEx: Stream<Item = ExExEvent> + Send + 'static {
+    /// Returns the name of the ExEx. It will appear in logs and metrics.
+    fn name(&self) -> Cow<'static, str>;
+}
+
+/// Events emitted by an ExEx.
+#[derive(Debug)]
+pub enum ExExEvent {
+    /// Highest block processed by the ExEx.
+    ///
+    /// ExEx must guarantee that it will not require all earlier blocks in the future, meaning
+    /// that Reth is allowed to prune them.
+    ///
+    /// On reorgs, it's possible for the height to go down.
+    FinishedHeight(BlockNumber),
+}
+
+/// Captures the context that an ExEx has access to.
+#[derive(Debug)]
+pub struct ExExContext<Node: FullNodeTypes> {
+    /// The current head of the blockchain at launch.
+    head: Head,
+    /// The configured provider to interact with the blockchain.
+    provider: Node::Provider,
+    /// The executor of the node.
+    executor: TaskExecutor,
+    /// The data dir of the node.
+    data_dir: ChainPath<DataDirPath>,
+    /// The config of the node
+    config: NodeConfig,
+    /// loaded config
+    reth_config: reth_config::Config,
+    // TODO(alexey): add pool, payload builder, anything else?
+}
+
+/// A trait for launching an ExEx.
+trait LaunchExEx<Node: FullNodeTypes>: Send {
+    /// Launches the ExEx.
+    ///
+    /// The ExEx should be able to run independently and emit events on the stream.
+    fn launch(self, ctx: ExExContext<Node>)
```

Comment:
I think this should borrow here, because we are potentially calling many exex
```suggestion
    fn launch(self, ctx: &ExExContext<Node>)
```

Reply:
the rationale was that ExEx will probably save many parts of the context into the internal state anyway, so we can clone from the beginning. Also, I think adding a reference here means that we need to introduce a non-static lifetime here https://github.com/paradigmxyz/reth/blob/0be95367f4e4b64717b91933b5d315c812106b8d/crates/node-builder/src/exex.rs#L73 and it gets messier in the signature

Reply:
point taken, fine with cloning, not that big of a deal, most expensive part is the Config, which we could arc eventually